storagenode/gracefulexit/worker.go

```diff
@@ -4,48 +4,248 @@
+			defer cancel()
+
+			// TODO what's the typical expiration setting?
+			pieceHash, err := worker.ecclient.PutPiece(putCtx, ctx, addrLimit, pk, reader, time.Now().Add(time.Second*600))
+			if err != nil {
+				worker.log.Error("failed to put piece.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+				// TODO look at error type to decide on the transfer error
+				worker.handleFailure(ctx, pb.TransferFailed_STORAGE_NODE_UNAVAILABLE, pieceID, satelliteID, c.Send)
+
+				continue
+			}
+
+			err = verifyPieceHash(ctx, addrLimit.GetLimit(), pieceHash, originalHash.Hash)
+			if err != nil {
+				worker.log.Error("failed hash verification.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+				worker.handleFailure(ctx, pb.TransferFailed_HASH_VERIFICATION, pieceID, satelliteID, c.Send)
+
+				continue
+			}
+
+			success := &pb.StorageNodeMessage{
+				Message: &pb.StorageNodeMessage_Succeeded{
+					Succeeded: &pb.TransferSucceeded{
+						OriginalPieceId:      msg.TransferPiece.OriginalPieceId,
+						OriginalPieceHash:    &pb.PieceHash{PieceId: msg.TransferPiece.OriginalPieceId},
+						OriginalOrderLimit:   &originalOrderLimit,
+						ReplacementPieceHash: pieceHash,
+					},
+				},
+			}
+			err = c.Send(success)
+			if err != nil {
+				return errs.Wrap(err)
+			}
+		case *pb.SatelliteMessage_DeletePiece:
+			pieceID := msg.DeletePiece.OriginalPieceId
+			err := worker.store.Delete(ctx, satelliteID, pieceID)
+			if err != nil {
+				worker.log.Error("failed to delete piece.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("piece ID", pieceID), zap.Error(errs.Wrap(err)))
+			}
+		case *pb.SatelliteMessage_ExitFailed:
+			worker.log.Error("graceful exit failed.", zap.Stringer("satellite ID", satelliteID), zap.Stringer("reason", msg.ExitFailed.Reason))
+
+			err = worker.satelliteDB.CompleteGracefulExit(ctx, satelliteID, time.Now(), satellites.ExitFailed, msg.ExitFailed.GetExitFailureSignature())
+			if err != nil {
+				return errs.Wrap(err)
+			}
+			break
+		case *pb.SatelliteMessage_ExitCompleted:
+			worker.log.Info("graceful exit completed.", zap.Stringer("satellite ID", satelliteID))
+
+			err = worker.satelliteDB.CompleteGracefulExit(ctx, satelliteID, time.Now(), satellites.ExitSucceeded, msg.ExitCompleted.GetExitCompleteSignature())
+			if err != nil {
+				return errs.Wrap(err)
+			}
+			break
+		default:
+			// TODO handle err
+			worker.log.Error("unknown graceful exit message.", zap.Stringer("satellite ID", satelliteID))
+		}
+
+	}
 
-	err = worker.satelliteDB.CompleteGracefulExit(ctx, satelliteID, time.Now(), satellites.ExitSucceeded, []byte{})
 	return errs.Wrap(err)
 }
 
+func (worker *Worker) handleFailure(ctx context.Context, transferError pb.TransferFailed_Error, pieceID pb.PieceID, satelliteID storj.NodeID, send func(*pb.StorageNodeMessage) error) {
+	failure := &pb.StorageNodeMessage{
```

Comment:
Maybe we don't need the `satelliteID` argument here since it's on the worker struct

Reply:
Updated